### PR TITLE
Tasks for the client channel/endpoint setup flow.

### PR DESCRIPTION
**Pull request by ababo**
This change includes:

1. ClientPreChannelCreate implementation and test.
2. ClientAfterChannelCreate implementation and test.
3. ClientPreEndpointMsgSOMCGet implementation and test.
4. ClientAfterEndpointMsgSOMCGet implementation and test.
5. Miscellaneous utilities and refactoring bits.

Partially resolves BV-292.

### CODE REVIEW

File: proc/worker/client.go [hunk 1]
```diff
@@ -1,8 +1,282 @@
...
+		return 0, err
+	}
+
+	amount, err := w.ethBack.PSCBalanceOf(&bind.CallOpts{}, addr)
+	if err != nil {
+		return 0, err
+	}
+
+	deposit := offer.UnitPrice*offer.MinUnits + offer.SetupPrice
+	if amount.Uint64() < deposit {
```
Comment (dzeckelev) [inline proc/worker/client.go]: it is better to convert the deposit to bigInt, this will exclude the border state when the amount is larger than Uint64

Comment (ababo) [reply proc/worker/client.go]: @sofabeat Is it practically possible?

File: proc/worker/client.go [hunk 2]
```diff
@@ -1,8 +1,282 @@
...
+	if err := data.Insert(w.db.Querier, &ch); err != nil {
+		return err
+	}
+
+	agentAddr, err := data.ToAddress(offer.Agent)
+	if err != nil {
+		return err
+	}
+
+	tx, err := w.ethBack.PSCCreateChannel(&bind.TransactOpts{},
```
Comment (dzeckelev) [inline proc/worker/client.go]: why bind.TransactOpts empty? So it should be?

File: proc/worker/client.go [hunk 3]
```diff
@@ -1,8 +1,282 @@
...
+		return nil, err
+	}
+
+	if !util.ValidateJSON(schema, mdata) {
+		return nil, fmt.Errorf(
+			"failed to validate endpoint for chan %s", ch.ID)
+	}
+
+	var msg ept.Message
+	json.Unmarshal(mdata, &msg)
```
Comment (dzeckelev) [inline proc/worker/client.go]: forgot to check error

Comment (ababo) [reply proc/worker/client.go]: Nope, I didn't. The message is validated before.

File: proc/worker/client.go (hunk 2 above)
Comment (ababo) [reply proc/worker/client.go]: Not being really familiar with this stuff I just followed the rest of the similar code which passes default options to the smart contract calls. If you have concrete suggestions to put here, I will modify that.

File: proc/worker/client_test.go [hunk 4]
```diff
@@ -1,29 +1,211 @@
...
+	defer env.close()
+
+	fxt := env.newTestFixture(t,
+		data.JobClientPreEndpointMsgSOMCGet, data.JobChannel)
+	defer fxt.Close()
+
+	swapAgentWithClient(t, fxt)
+
+	msg := ept.Message{
+		TemplateHash:           "test-hash",
```
Comment (dzeckelev) [inline proc/worker/client_test.go]: It may be better to make these fields in constant?

Comment (ababo) [reply proc/worker/client_test.go]: No, it probably won't, since it's not reused.

File: proc/worker/client.go (hunk 2 above)
Comment (dzeckelev) [reply proc/worker/client.go]: add From and Context , it would be enough.

File: proc/worker/client.go (hunk 1 above)
Comment (drew2a) [reply proc/worker/client.go]: Total supply of PRIX is `1275455`.
As I know we store PRIX as an integer and supports 8 decimals.
So, max integer will be `1275455*100000000 `

File: proc/worker/client.go [hunk 5]
```diff
@@ -1,8 +1,282 @@
...
+		return err
+	}
+
+	tx, err := w.ethBack.PSCCreateChannel(&bind.TransactOpts{},
+		agentAddr, offerHash, big.NewInt(int64(deposit)))
+	if err != nil {
+		return err
+	}
+
+	return w.saveEthTX(job, tx, "CreateChannel",
```
Comment (dzeckelev) [inline proc/worker/client.go]: if this function returns an error, what will happen to the channel that you saved in the database above?

Comment (ababo) [reply proc/worker/client.go]: Good point! Updated.

File: proc/worker/client.go (hunk 2 above)
Comment (furkhat) [reply proc/worker/client.go]: Use `bind.NewKeyedTransactor(agentPrivKet)`, where `agentPrivKey` is `*ecdsa.PrivateKey`

File: proc/worker/client.go [hunk 6]
```diff
@@ -1,8 +1,286 @@
...
+
+func (w *Worker) clientPreChannelCreateSaveTX(
+	job *data.Job, acc *data.Account, offer *data.Offering,
+	offerHash common.Hash, deposit uint64) error {
+	agentAddr, err := data.ToAddress(offer.Agent)
+	if err != nil {
+		return err
+	}
+
+	tx, err := w.ethBack.PSCCreateChannel(&bind.TransactOpts{},
```
Comment (furkhat) [inline proc/worker/client.go]: Add `// TODO: set gas limit from conf`

**Pull request merged**